### ISSUE TITLE
chaiscript::constructor fails to compile when used on noncopyable class

### ISSUE THREAD

**Issue by Tw1ddle** [ChaiScript/ChaiScript#519]
* Compiler Used: gcc (Ubuntu 9.2.1-9ubuntu2) 9.2.1 20191008
 * Operating System: Linux
 * Architecture: x86 64-bit
 
chaiscript::constructor gives a compile error when given a noncopyable class (see below) on the develop branch (50ca037da95399a3f88b416288e8d111caa24e32).

Issue appears to have been introduced at c19705d, reverting changes to files like proxy_constructors.hpp from this commit eventually gets it compiling.

### Minimal Example to Reproduce Behavior

```
#include "chaiscript/chaiscript.hpp"

class MyClass
{
public:
    MyClass(int) {}
    MyClass(const MyClass&) = delete; // Deleted copy constructor
};

int main()
{
    chaiscript::ChaiScript chai;
    chai.add(chaiscript::constructor<MyClass(int)>(), "MyClass"); // Error here
    return 0;
}
```

Compiler output:

> In file included from /usr/include/x86_64-linux-gnu/c++/9/bits/c++allocator.h:33,
>                  from /usr/include/c++/9/bits/allocator.h:46,
>                  from /usr/include/c++/9/vector:64,
>                  from ../untitled/chaiscript/chaiscript_defines.hpp:23,
>                  from ../untitled/chaiscript/chaiscript_basic.hpp:10,
>                  from ../untitled/chaiscript/chaiscript.hpp:823,
>                  from ../untitled/main.cpp:1:
> /usr/include/c++/9/ext/new_allocator.h: In instantiation of ‘void __gnu_cxx::new_allocator<_Tp>::construct(_Up*, _Args&& ...) [with _Up = MyClass; _Args = {MyClass}; _Tp = MyClass]’:
> /usr/include/c++/9/bits/alloc_traits.h:482:2:   required from ‘static void std::allocator_traits<std::allocator<_Tp1> >::construct(std::allocator_traits<std::allocator<_Tp1> >::allocator_type&, _Up*, _Args&& ...) [with _Up = MyClass; _Args = {MyClass}; _Tp = MyClass; std::allocator_traits<std::allocator<_Tp1> >::allocator_type = std::allocator<MyClass>]’
> /usr/include/c++/9/bits/shared_ptr_base.h:548:39:   required from ‘std::_Sp_counted_ptr_inplace<_Tp, _Alloc, _Lp>::_Sp_counted_ptr_inplace(_Alloc, _Args&& ...) [with _Args = {MyClass}; _Tp = MyClass; _Alloc = std::allocator<MyClass>; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
> /usr/include/c++/9/bits/shared_ptr_base.h:679:16:   required from ‘std::__shared_count<_Lp>::__shared_count(_Tp*&, std::_Sp_alloc_shared_tag<_Alloc>, _Args&& ...) [with _Tp = MyClass; _Alloc = std::allocator<MyClass>; _Args = {MyClass}; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
> /usr/include/c++/9/bits/shared_ptr_base.h:1344:71:   required from ‘std::__shared_ptr<_Tp, _Lp>::__shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<MyClass>; _Args = {MyClass}; _Tp = MyClass; __gnu_cxx::_Lock_policy _Lp = __gnu_cxx::_S_atomic]’
> /usr/include/c++/9/bits/shared_ptr.h:359:59:   required from ‘std::shared_ptr<_Tp>::shared_ptr(std::_Sp_alloc_shared_tag<_Tp>, _Args&& ...) [with _Alloc = std::allocator<MyClass>; _Args = {MyClass}; _Tp = MyClass]’
> /usr/include/c++/9/bits/shared_ptr.h:701:14:   required from ‘std::shared_ptr<_Tp> std::allocate_shared(const _Alloc&, _Args&& ...) [with _Tp = MyClass; _Alloc = std::allocator<MyClass>; _Args = {MyClass}]’
> /usr/include/c++/9/bits/shared_ptr.h:717:39:   required from ‘std::shared_ptr<_Tp> std::make_shared(_Args&& ...) [with _Tp = MyClass; _Args = {MyClass}]’
> ../untitled/chaiscript/dispatchkit/handle_return.hpp:49:51:   required from ‘static chaiscript::Boxed_Value chaiscript::dispatch::detail::Handle_Return<Ret>::handle(T&&) [with T = MyClass; <template-parameter-2-2> = void; Ret = MyClass]’
> ../untitled/chaiscript/dispatchkit/proxy_functions_detail.hpp:115:46:   required from ‘chaiscript::Boxed_Value chaiscript::dispatch::detail::call_func(Ret (*)(Params ...), const Callable&, const chaiscript::Function_Params&, const chaiscript::Type_Conversions_State&) [with Callable = chaiscript::dispatch::detail::build_constructor_(Class (*)(Params ...)) [with Class = MyClass; Params = {int}; chaiscript::Proxy_Function = std::shared_ptr<chaiscript::dispatch::Proxy_Function_Base>]::<lambda(auto:4&& ...)>; Ret = MyClass; Params = {int}]’
> ../untitled/chaiscript/dispatchkit/proxy_functions.hpp:667:35:   required from ‘chaiscript::Boxed_Value chaiscript::dispatch::Proxy_Function_Callable_Impl<T, U>::do_call(const chaiscript::Function_Params&, const chaiscript::Type_Conversions_State&) const [with Func = MyClass(int); Callable = chaiscript::dispatch::detail::build_constructor_(Class (*)(Params ...)) [with Class = MyClass; Params = {int}; chaiscript::Proxy_Function = std::shared_ptr<chaiscript::dispatch::Proxy_Function_Base>]::<lambda(auto:4&& ...)>]’
> ../untitled/chaiscript/dispatchkit/proxy_functions.hpp:665:21:   required from here
> /usr/include/c++/9/ext/new_allocator.h:145:20: error: use of deleted function ‘MyClass::MyClass(const MyClass&)’
>   145 |  noexcept(noexcept(::new((void *)__p)
>       |                    ^~~~~~~~~~~~~~~~~~
>   146 |        _Up(std::forward<_Args>(__args)...)))
>       |        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> ../untitled/main.cpp:7:5: note: declared here
>     7 |     MyClass(const MyClass&) = delete; // Deleted copy constructor
>       |     ^~~~~~~

**Comment by totalgee** [comment]
Not sure if this is related (a similar failure to compile). I'm trying to build with the latest on develop, and on VS 2019 (in C++17 mode) I can't even merely **include** "chaiscript.hpp" without getting compile errors.

```
proxy_functions.hpp(80,51): error C2131: expression did not evaluate to a constant
message : a non-constant (sub-)expression was encountered
```

This is related to all the `constexpr` stuff, all cases where it's trying to get `user_type<Dynamic_Object>()`.

**Comment by totalgee** [comment]
Regarding my issue, I had to merge in the stuff locally from @lefticus branch [VS_2019_fixes](https://github.com/ChaiScript/ChaiScript/tree/VS_2019_fixes) in order to fix the compilation problems (he replaced `constexpr` with `const` a few places, plus there were some other problems he addressed).

Even then, that branch didn't solve all my compilation problems -- I still had to rename `chaiscript::dispatch::detail::Function_Params` to be `Function_Params_t` (or something different), or alternatively in a few places refer explicitly to `chaiscript::Function_Params` to disambiguate between these two types (e.g. in the "detail" helper functions in `register_function.hpp`)... Only then I was able to get things compiling on VS2019.

@RobLoach I'd recommend Jason's branch be merged in to `develop`, unless it causes issues on Linux or somewhere else -- otherwise, the current (develop) branch of ChaiScript really isn't usable with VS2019.

**Comment by totalgee** [comment]
(Sorry for my above comments, in the end they really aren't related to your issue, but rather to building on VS2019 -- and are now addressed by #539).

However, I did open a similar issue #537 (sorry, I see it's basically a duplicate of this one), where I also give my "workaround", registering a lambda that returns a new `shared_ptr` of the desired type (rather than using the `constructor()` template function). Perhaps a new registration helper could be made to do something like my workaround for non-copiable classes (which is essentially how the internal function `build_constructor_()` used to work)?

BTW, did you test to see if it works when your registered class has a move constructor and move assignment operator?
